README
#
# $Id$
#

The WiFi Guard Dog project is a complete and embeddable captive portal
solution for wireless community groups or individuals who wish to open a
free HotSpot while still preventing abuse of their Internet connection.

The original project's homepage is:
	http://dev.wifidog.org/

The WifiAS's homepage is http://wifias.com/

The project's software is released under the GPL license and is copyright it's respective owners.

When you compile it in openwrt,you need to add the code below in the ./feeds/packages/net/wifidog/Makefile after:
$(INSTALL_DATA) $(PKG_BUILD_DIR)/wifidog-msg.html $(1)/etc/


$(INSTALL_DATA) $(PKG_BUILD_DIR)/src/luci/controller/wifidog.lua $(1)/usr/lib/lua/luci/controller/
$(INSTALL_DATA) $(PKG_BUILD_DIR)/src/luci/model/cbi/wifidog.lua $(1)/usr/lib/lua/luci/model/cbi/
$(INSTALL_DATA) $(PKG_BUILD_DIR)/src/luci/wifidog $(1)/etc/config/
$(INSTALL_DATA) $(PKG_BUILD_DIR)/wifidog.conf.default $(1)/etc/

change the /etc/init.d/wifidog in you feeds files
start function

add trusted MAC and host in default configuration
TrustedMACList '00:22:11:A4:BC:88,00:22:11:A4:BC:89'
TrustedHostList www.eachsea.com,eachsea.com,wifias.com,auth.wifias.com,user.wifias.com,member.wifias.com

ChangeLog
# $Id$
2014-07-03
        * add clientip in the redirect url
2014-05-13
        * libhttpd crash on invalid HTTP headers (second part of the patch) by Benoit Grégoire
2013-08-21
        * add support for DROP target for firewall rules by Champtar
2013-06-14
	* add support for log, ulog target for firewall rules by jean-philippe menil and florida 
2013-05-31        
        Add transparent proxy support (via iptables REDIRECT) by Champtar (inspired by FFW team) 
2012-05-30
	* Add many const by champtar
	* Send http 302 instead of 307 by champtar
	* Suppress all compilation warnings by champtar
	* Add transparent proxy support (via iptables REDIRECT) 
2012-08-28
	* Fix #836, buffer overflow on long urls reported by Etienne CHAMPETIER
	* Fix #835, segfaults reported by Etienne CHAMPETIER
2009-11-03
	* Fix #625, does not display failure notice when quiet is set to true
	* Fix #587, change index and rindex to strchr and strrchr
	* Fix #548, trim leading spaces of the config file's options

2009-09-28 Benoit Grégoire  <benoitg@coeus.ca>
	* Fix #471, patch by  wichert

2009-09-25 Geneviève Bastien <gbastien@versatic.net>
	* Release 20090925
	* Update contrib Makefiles

2009-09-17 Geneviève Bastien <gbastien@versatic.net>
	* Documented #537
	* Fixed #472, patch by Jean-Philippe Menil
	* Fixed #515, using the gateway interface instead of the gateway id in the iptables chain

2009-07-02 Benoit Grégoire  <bock@step.polymtl.ca>
	* Re-fix #505, #525 and fix #584, sorry about that.

2009-06-26 Benoit Grégoire  <bock@step.polymtl.ca>
	* Fix #518

2009-02-27 Benoit Grégoire  <bock@step.polymtl.ca>
	* Fix #488 and #493 (arp_get() in firewall.c couldn't parse lowercase mac's from /proc/net/arp) with patch from	jch@pps.jussieu.fr.  Otherwise wifidog wouldn't work with recent openwrt and Ubuntu.
	* Fix #525
	
2008-09-30 Wichert Akkerman <wichert@wiggy.net>
	* Add exitcode to iptables failure errors.
	* Include the gw_id in auth server updates so the client does not have
	  to keep track of it in a session.
	* Include the gateway id in the firewall table names. Fixes ticket #466
	* URL encode the token before transmitting (it was already decoded).
	  Fixes ticket #473
	* Clean up compiler warnings.
	* Security: strncpy may not NUL-terminate strings, so enforce this
	  ourselves. Fixes ticket #464
	* Make it possible to protect the status page. Fixes ticket #463.

2008-07-20 Alexandre Carmel-Veilleux <acv@miniguru.ca>
	* src/util.c: Fixed #include bug that caused segfaults on newer Linux

2008-04-21 Alexandre Carmel-Veilleux <acv@miniguru.ca>
	* Integrated patch #452 from Wichert Akkerman <wichert@wiggy.net>: Add const to function arguments in libhttpd to enforce more type checking and prevent certain class of problems.
	* Compatiblity fix: Libhttpd assumes that type u_int is defined. Added an #ifndef/#include <sys/types.h> pair to httpd.h to make sure that assertion is true.
	* Integrated patch #453 from Wichert Akkerman <wichert@wiggy.net>: Add configurable html to wifidog error messages. This has been a long-requested feature.

2008-04-13 Benoit Grégoire  <bock@step.polymtl.ca>
	* contrib/build-openwrt-kamikazeipk/wifidog/Makefile:  Add iptables userspace dependencies
	* Release 1.1.5
	
2008-03-24 Benoit Grégoire  <bock@step.polymtl.ca>
	* Integrate with OpenWRT kamikaze build system

2007-11-01 Benoit Grégoire  <bock@step.polymtl.ca>
	* Apply portability patches by David Young <dyoung@pobox.com>.  These have been reviewed, but not tested.

2007-10-18 Benoit Grégoire  <bock@step.polymtl.ca>
	* fw_iptables.c: From Philippe April:  reverted change made in 1241 so we properly remove the entry from mangle.WiFiDog_Incoming when kicking out users, it was affecting statistics 
	* Update doxygen.cfg.in for latest version and to fix path ambiguity during make dist.
	* Release 1.1.4

2007-07-06 Benoit Grégoire  <bock@step.polymtl.ca>
	* Makefile.am:  Slight change in make ipk tagrget.  It seems that sometimes builddir isn't defined.  srcdir works just as well in this case.
	
2007-06-27 Benoit Grégoire  <bock@step.polymtl.ca>
	* util.c:  Fix while loop initialisation bug
	* conf.h:  Forgot to change the value of NUM_EXT_INTERFACE_DETECT_RETRY to actually make it wait forever.
	* Remove hardcoded authserver paths.  Can now be defined in the config file (auth server section).
	* Centralise browser redirect code to simplify code
	* Add manual logout URL, based in part on work by David Bird
	* Release 1.1.3 final
	
2007-06-24 Benoit Grégoire  <bock@step.polymtl.ca>
	* Close #321:  Make the Gateway retry forever if it cannot find it's interface.  You never know when someone may finally replug the network cable or something...
	* Close #332:  Apply patch from Laurent Marchal. biguphpc<AT>gmail<DOT>com
	* fw_iptables.c:  Fix error in iptables_fw_access().  Rules were created as ACCEPT instead of DROP
	* firewall.c:  Fix bug in fw_sync_with_authserver().  The traffic for the validation period of a user who validated his account while connected wouldn't get counted.
	* doc/wifidog_firewall_map.dia:  At long last, full documentation of the firewall.  We would have avoided a lot of stupid mistakes if we produced that sooner.
	* Release 1.1.3_rc1

2007-05-24 Benoit Grégoire  <bock@step.polymtl.ca>
	* wdctl_thread.c:  Fix #324, again.  Credit goes to Medea, I misunderstood his instructons.
	* From David Bird <david@coova.com> libhttpd/: Fix #266 - don't process query string parameters and keep them in that request.path.

2007-05-18 Benoit Grégoire  <bock@step.polymtl.ca>
	* wdctl_thread.c:  Fix #324
	
2007-04-26 Benoit Grégoire  <bock@step.polymtl.ca>
	* wifidog.conf:  Improve comments and add examples of blocking access to the upstream LAN.

2007-04-26 Benoit Grégoire  <bock@step.polymtl.ca>
	* conf.h:  The DEFAULT_CHECKINTERVAL was 5 instead of 60 (as stated in the config file) which caused huge needless load on the auth servers, and needless ping traffic towards the clients if it wasn't manually set.


src/centralserver.c
/********************************************************************\
 * This program is free software; you can redistribute it and/or    *
 * modify it under the terms of the GNU General Public License as   *
 * published by the Free Software Foundation; either version 2 of   *
 * the License, or (at your option) any later version.              *
 *                                                                  *
 * This program is distributed in the hope that it will be useful,  *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
 * GNU General Public License for more details.                     *
 *                                                                  *
 * You should have received a copy of the GNU General Public License*
 * along with this program; if not, contact:                        *
 *                                                                  *
 * Free Software Foundation           Voice:  +1-617-542-5942       *
 * 59 Temple Place - Suite 330        Fax:    +1-617-542-2652       *
 * Boston, MA  02111-1307,  USA       gnu@gnu.org                   *
 *                                                                  *
 \********************************************************************/

/* $Id$ */
/** @file centralserver.c
  @brief Functions to talk to the central server (auth/send stats/get rules/etc...)
  @author Copyright (C) 2004 Philippe April <papril777@yahoo.com>
 */

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <unistd.h>
#include <string.h>
#include <syslog.h>

#include "httpd.h"

#include "common.h"
#include "safe.h"
#include "util.h"
#include "auth.h"
#include "conf.h"
#include "debug.h"
#include "centralserver.h"
#include "firewall.h"
#include "../config.h"

extern pthread_mutex_t	config_mutex;

/** Initiates a transaction with the auth server, either to authenticate or to
 * update the traffic counters at the server
@param authresponse Returns the information given by the central server 
@param request_type Use the REQUEST_TYPE_* defines in centralserver.h
@param ip IP adress of the client this request is related to
@param mac MAC adress of the client this request is related to
@param token Authentification token of the client
@param incoming Current counter of the client's total incoming traffic, in bytes 
@param outgoing Current counter of the client's total outgoing traffic, in bytes 
*/
t_authcode
auth_server_request(t_authresponse *authresponse, const char *request_type, const char *ip, const char *mac, const char *token, unsigned long long int incoming, unsigned long long int outgoing)
{
	int sockfd;
	ssize_t	numbytes;
	size_t totalbytes;
	char buf[MAX_BUF];
	char *tmp;
        char *safe_token;
	int done, nfds;
	fd_set			readfds;
	struct timeval		timeout;
	t_auth_serv	*auth_server = NULL;
	auth_server = get_auth_server();
	
	/* Blanket default is error. */
	authresponse->authcode = AUTH_ERROR;
	
	sockfd = connect_auth_server();
	if (sockfd == -1) {
		/* Could not connect to any auth server */
		return (AUTH_ERROR);
	}

	/**
	 * TODO: XXX change the PHP so we can harmonize stage as request_type
	 * everywhere.
	 */
	memset(buf, 0, sizeof(buf));
        safe_token=httpdUrlEncode(token);
	snprintf(buf, (sizeof(buf) - 1),
		"GET %s%sstage=%s&ip=%s&mac=%s&token=%s&incoming=%llu&outgoing=%llu&gw_id=%s HTTP/1.0\r\n"
		"User-Agent: WiFiDog %s\r\n"
		"Host: %s\r\n"
		"\r\n",
		auth_server->authserv_path,
		auth_server->authserv_auth_script_path_fragment,
		request_type,
		ip,
		mac,
		safe_token,
		incoming,
		outgoing,
                config_get_config()->gw_id,
		VERSION,
		auth_server->authserv_hostname
	);

        free(safe_token);

	debug(LOG_DEBUG, "Sending HTTP request to auth server: [%s]\n", buf);
	send(sockfd, buf, strlen(buf), 0);

	debug(LOG_DEBUG, "Reading response");
	numbytes = totalbytes = 0;
	done = 0;
	do {
		FD_ZERO(&readfds);
		FD_SET(sockfd, &readfds);
		timeout.tv_sec = 30; /* XXX magic... 30 second is as good a timeout as any */
		timeout.tv_usec = 0;
		nfds = sockfd + 1;

		nfds = select(nfds, &readfds, NULL, NULL, &timeout);

		if (nfds > 0) {
			/** We don't have to use FD_ISSET() because there
			 *  was only one fd. */
			numbytes = read(sockfd, buf + totalbytes, MAX_BUF - (totalbytes + 1));
			if (numbytes < 0) {
				debug(LOG_ERR, "An error occurred while reading from auth server: %s", strerror(errno));
				/* FIXME */
				close(sockfd);
				return (AUTH_ERROR);
			}
			else if (numbytes == 0) {
				done = 1;
			}
			else {
				totalbytes += numbytes;
				debug(LOG_DEBUG, "Read %d bytes, total now %d", numbytes, totalbytes);
			}
		}
		else if (nfds == 0) {
			debug(LOG_ERR, "Timed out reading data via select() from auth server");
			/* FIXME */
			close(sockfd);
			return (AUTH_ERROR);
		}
		else if (nfds < 0) {
			debug(LOG_ERR, "Error reading data via select() from auth server: %s", strerror(errno));
			/* FIXME */
			close(sockfd);
			return (AUTH_ERROR);
		}
	} while (!done);

	close(sockfd);

	buf[totalbytes] = '\0';
	debug(LOG_DEBUG, "HTTP Response from Server: [%s]", buf);
	
	if ((tmp = strstr(buf, "Auth: "))) {
		if (sscanf(tmp, "Auth: %d", (int *)&authresponse->authcode) == 1) {
			debug(LOG_INFO, "Auth server returned authentication code %d", authresponse->authcode);
			return(authresponse->authcode);
		} else {
			debug(LOG_WARNING, "Auth server did not return expected authentication code");
			return(AUTH_ERROR);
		}
	}
	else {
		return(AUTH_ERROR);
	}

	/* XXX Never reached because of the above if()/else pair. */
	return(AUTH_ERROR);
}

/* Tries really hard to connect to an auth server. Returns a file descriptor, -1 on error
 */
int connect_auth_server() {
	int sockfd;

	LOCK_CONFIG();
	sockfd = _connect_auth_server(0);
	UNLOCK_CONFIG();

	if (sockfd == -1) {
		debug(LOG_ERR, "Failed to connect to any of the auth servers");
		mark_auth_offline();
	}
	else {
		debug(LOG_DEBUG, "Connected to auth server");
		mark_auth_online();
	}
	return (sockfd);
}

/* Helper function called by connect_auth_server() to do the actual work including recursion
 * DO NOT CALL DIRECTLY
 @param level recursion level indicator must be 0 when not called by _connect_auth_server()
 */
int _connect_auth_server(int level) {
	s_config *config = config_get_config();
	t_auth_serv *auth_server = NULL;
	struct in_addr *h_addr;
	int num_servers = 0;
	char * hostname = NULL;
	char * popular_servers[] = {
		  "www.baidu.com",
		  "www.taobao.com",
		  "www.qq.com",
		  "www.weibo.com",
		  NULL
	};
	char ** popularserver;
	char * ip;
	struct sockaddr_in their_addr;
	int sockfd;

	/* XXX level starts out at 0 and gets incremented by every iterations. */
	level++;

	/*
	 * Let's calculate the number of servers we have
	 */
	for (auth_server = config->auth_servers; auth_server; auth_server = auth_server->next) {
		num_servers++;
	}
	debug(LOG_DEBUG, "Level %d: Calculated %d auth servers in list", level, num_servers);

	if (level > num_servers) {
		/*
		 * We've called ourselves too many times
		 * This means we've cycled through all the servers in the server list
		 * at least once and none are accessible
		 */
		return (-1);
	}

	/*
	 * Let's resolve the hostname of the top server to an IP address
	 */
	auth_server = config->auth_servers;
	hostname = auth_server->authserv_hostname;
	debug(LOG_DEBUG, "Level %d: Resolving auth server [%s]", level, hostname);
	h_addr = wd_gethostbyname(hostname);
	if (!h_addr) {
		/*
		 * DNS resolving it failed
		 *
		 * Can we resolve any of the popular servers ?
		 */
		debug(LOG_DEBUG, "Level %d: Resolving auth server [%s] failed", level, hostname);

		for (popularserver = popular_servers; *popularserver; popularserver++) {
			debug(LOG_DEBUG, "Level %d: Resolving popular server [%s]", level, *popularserver);
			h_addr = wd_gethostbyname(*popularserver);
			if (h_addr) {
				debug(LOG_DEBUG, "Level %d: Resolving popular server [%s] succeeded = [%s]", level, *popularserver, inet_ntoa(*h_addr));
				break;
			}
			else {
				debug(LOG_DEBUG, "Level %d: Resolving popular server [%s] failed", level, *popularserver);
			}
		}

		/* 
		 * If we got any h_addr buffer for one of the popular servers, in other
		 * words, if one of the popular servers resolved, we'll assume the DNS
		 * works, otherwise we'll deal with net connection or DNS failure.
		 */
		if (h_addr) {
			free (h_addr);
			/*
			 * Yes
			 *
			 * The auth server's DNS server is probably dead. Try the next auth server
			 */
			debug(LOG_DEBUG, "Level %d: Marking auth server [%s] as bad and trying next if possible", level, hostname);
			if (auth_server->last_ip) {
				free(auth_server->last_ip);
				auth_server->last_ip = NULL;
			}
			mark_auth_server_bad(auth_server);
			return _connect_auth_server(level);
		}
		else {
			/*
			 * No
			 *
			 * It's probably safe to assume that the internet connection is malfunctioning
			 * and nothing we can do will make it work
			 */
			mark_offline();
			debug(LOG_DEBUG, "Level %d: Failed to resolve auth server and all popular servers. "
					"The internet connection is probably down", level);
			return(-1);
		}
	}
	else {
		/*
		 * DNS resolving was successful
		 */
		ip = safe_strdup(inet_ntoa(*h_addr));
		debug(LOG_DEBUG, "Level %d: Resolving auth server [%s] succeeded = [%s]", level, hostname, ip);

		if (!auth_server->last_ip || strcmp(auth_server->last_ip, ip) != 0) {
			/*
			 * But the IP address is different from the last one we knew
			 * Update it
			 */
			debug(LOG_DEBUG, "Level %d: Updating last_ip IP of server [%s] to [%s]", level, hostname, ip);
			if (auth_server->last_ip) free(auth_server->last_ip);
			auth_server->last_ip = ip;

			/* Update firewall rules */
			fw_clear_authservers();
			fw_set_authservers();
		}
		else {
			/*
			 * IP is the same as last time
			 */
			free(ip);
		}

		/*
		 * Connect to it
		 */
		debug(LOG_DEBUG, "Level %d: Connecting to auth server %s:%d", level, hostname, auth_server->authserv_http_port);
		their_addr.sin_family = AF_INET;
		their_addr.sin_port = htons(auth_server->authserv_http_port);
		their_addr.sin_addr = *h_addr;
		memset(&(their_addr.sin_zero), '\0', sizeof(their_addr.sin_zero));
		free (h_addr);

		if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
			debug(LOG_ERR, "Level %d: Failed to create a new SOCK_STREAM socket: %s", strerror(errno));
			return(-1);
		}

		if (connect(sockfd, (struct sockaddr *)&their_addr, sizeof(struct sockaddr)) == -1) {
			/*
			 * Failed to connect
			 * Mark the server as bad and try the next one
			 */
			debug(LOG_DEBUG, "Level %d: Failed to connect to auth server %s:%d (%s). Marking it as bad and trying next if possible", level, hostname, auth_server->authserv_http_port, strerror(errno));
			close(sockfd);
			mark_auth_server_bad(auth_server);
			return _connect_auth_server(level); /* Yay recursion! */
		}
		else {
			/*
			 * We have successfully connected
			 */
			debug(LOG_DEBUG, "Level %d: Successfully connected to auth server %s:%d", level, hostname, auth_server->authserv_http_port);
			return sockfd;
		}
	}
}


src/luci/model/cbi/wifidog.lua
--[[
live-8.taobao.com
]]--

local sys = require "luci.sys"
local fs = require "nixio.fs"
local uci = require "luci.model.uci".cursor()

--Willing
local running=(luci.sys.call("pidof wifidog > /dev/null") == 0)
local button=""

if running then
	m = Map("wifidog", translate("wifidog-web"), translate("wifidog：running........."))
else
	m = Map("wifidog", translate("wifidog-web"), translate("wifidog not start"))
end

if fs.access("/usr/bin/wifidog") then
	s = m:section(TypedSection, "wifidog", translate("wifidog config"))
	s.anonymous = true
	s.addremove = false

	wifi_enable = s:option(Flag, "wifidog_enable", translate("run wifidog"),translate("run or shutdown wifidog"))
	wifi_enable.default = 0

	
	gatewayID = s:option(Value,"gateway_id",translate("GatewayID"),translate("id（GatewayID）,etc:demoNode"))
	gateway_interface = s:option(Value,"gateway_interface",translate("lan"),translate("br-lan"))
	gateway_eninterface = s:option(Value,"gateway_eninterface",translate("wan"),translate("eth1/eth2"))
	gateway_hostname = s:option(Value,"gateway_hostname",translate("auth host"),translate("etc:auth.wifias.com"))
	gateway_httpport = s:option(Value,"gateway_httpport",translate("server prot"),translate("etc:80"))
	gateway_path = s:option(Value,"gateway_path",translate("url"),translate("/demoToken/"))
	gateway_connmax = s:option(Value,"HTTPDMaxConn",translate("max user"))
	client_timeout = s:option(Value,"client_timeout",translate("client timeout"))
	check_interval = s:option(Value,"check_interval",translate("check interval"))
	TrustedHostList = s:option(Value,"TrustedHostList",translate("Trusted Host List"),translate("etc:www.wifias.com,user.wifias.com"))
	TrustedMACList = s:option(Value,"TrustedMACList",translate("Trusted MAC List"),translate("etc:00:22:11:A4:BC:88,00:22:11:A4:BC:89"))

	gatewayID = s:option(Value,"gateway_id",translate("GatewayID"),translate("id（GatewayID）,etc:demoNode"))
	gateway_interface = s:option(Value,"gateway_interface",translate("lan"),translate("br-lan"))
	gateway_eninterface = s:option(Value,"gateway_eninterface",translate("wan"),translate("eth1/eth2"))
	gateway_hostname = s:option(Value,"gateway_hostname",translate("auth host"),translate("etc:auth.wifias.com"))
	gateway_httpport = s:option(Value,"gateway_httpport",translate("server prot"),translate("etc:80"))
	gateway_path = s:option(Value,"gateway_path",translate("url"),translate("/demoToken/"))
	gateway_connmax = s:option(Value,"HTTPDMaxConn",translate("max user"))
	client_timeout = s:option(Value,"client_timeout",translate("client timeout"))
	check_interval = s:option(Value,"check_interval",translate("check interval"))
	TrustedHostList = s:option(Value,"TrustedHostList",translate("Trusted Host List"),translate("etc:www.wifias.com,user.wifias.com"))
	TrustedMACList = s:option(Value,"TrustedMACList",translate("Trusted MAC List"),translate("etc:00:22:11:A4:BC:88,00:22:11:A4:BC:89"))

	
	--deamo_enable = s:option(Flag, "enable", translate("Dog"),translate("reboot wifidog"))
	--deamo_enable:depends("wifidog_enable","1")
	--deamo_enable.default = 0
else
	m.pageaction = false
end

return m



src/http.c
/********************************************************************\
 * This program is free software; you can redistribute it and/or    *
 * modify it under the terms of the GNU General Public License as   *
 * published by the Free Software Foundation; either version 2 of   *
 * the License, or (at your option) any later version.              *
 *                                                                  *
 * This program is distributed in the hope that it will be useful,  *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
 * GNU General Public License for more details.                     *
 *                                                                  *
 * You should have received a copy of the GNU General Public License*
 * along with this program; if not, contact:                        *
 *                                                                  *
 * Free Software Foundation           Voice:  +1-617-542-5942       *
 * 59 Temple Place - Suite 330        Fax:    +1-617-542-2652       *
 * Boston, MA  02111-1307,  USA       gnu@gnu.org                   *
 *                                                                  *
 \********************************************************************/

/* $Id$ */
/** @file http.c
  @brief HTTP IO functions
  @author Copyright (C) 2004 Philippe April <papril777@yahoo.com>
  @author Copyright (C) 2007 Benoit Grégoire
  @author Copyright (C) 2007 David Bird <david@coova.com>

 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>

#include "httpd.h"

#include "safe.h"
#include "debug.h"
#include "conf.h"
#include "auth.h"
#include "firewall.h"
#include "http.h"
#include "httpd.h"
#include "client_list.h"
#include "common.h"
#include "centralserver.h"

#include "util.h"

#include "../config.h"

extern pthread_mutex_t	client_list_mutex;

/** The 404 handler is also responsible for redirecting to the auth server */
void
http_callback_404(httpd *webserver, request *r)
{
	char tmp_url[MAX_BUF],
			*url,
			*mac;
	s_config	*config = config_get_config();
	t_auth_serv	*auth_server = get_auth_server();

	memset(tmp_url, 0, sizeof(tmp_url));
	/* 
	 * XXX Note the code below assumes that the client's request is a plain
	 * http request to a standard port. At any rate, this handler is called only
	 * if the internet/auth server is down so it's not a huge loss, but still.
	 */
        snprintf(tmp_url, (sizeof(tmp_url) - 1), "http://%s%s%s%s",
                        r->request.host,
                        r->request.path,
                        r->request.query[0] ? "?" : "",
                        r->request.query);
	url = httpdUrlEncode(tmp_url);

	if (!is_online()) {
		/* The internet connection is down at the moment  - apologize and do not redirect anywhere */
		char * buf;
		safe_asprintf(&buf, 
			"<p>We apologize, but it seems that the internet connection that powers this hotspot is temporarily unavailable.</p>"
			"<p>If at all possible, please notify the owners of this hotspot that the internet connection is out of service.</p>"
			"<p>The maintainers of this network are aware of this disruption.  We hope that this situation will be resolved soon.</p>"
			"<p>In a while please <a href='%s'>click here</a> to try your request again.</p><p style='display:none;'>中文测试</p>", tmp_url);
		/*
		safe_asprintf(&buf, 
			"<p>很抱歉, 您所连接的认证网关暂时无法连接到互联网.</p>"
			"<p>如果一直无法连接,并显示此画面, 请告知认证网关提供者检查认证网关的互联网连接状态.</p>"
			"<p>维护人员如果知道这个故障.  希望可以尽快解决此问题.</p>"
			"<p>稍后请<a href='%s'>点击这里</a>尝试重新连接.</p>", tmp_url);
			*/

                send_http_page(r, "Uh oh! Internet access unavailable!", buf);
                /*send_http_page(r, "很抱歉! 认证网关无法连接待互联网!", buf);*/
		free(buf);
		debug(LOG_INFO, "Sent %s an apology since I am not online - no point sending them to auth server", r->clientAddr);
	}
	else if (!is_auth_online()) {
		/* The auth server is down at the moment - apologize and do not redirect anywhere */
		char * buf;safe_asprintf(&buf, 
			"<p>We apologize, but it seems that we are currently unable to re-direct you to the login screen.</p>"
			"<p>The maintainers of this network are aware of this disruption.  We hope that this situation will be resolved soon.</p>"
			"<p>In a couple of minutes please <a href='%s'>click here</a> to try your request again.</p>", tmp_url);
		/*
		safe_asprintf(&buf, 
			"<p>很抱歉, 暂时无法跳转到登陆页面.</p>"
			"<p>维护人员如果知道这个故障.  希望可以尽快解决此问题.</p>"
			"<p>稍后请<a href='%s'>点击这里</a>尝试重新连接.</p>", tmp_url);
			*/

                send_http_page(r, "Uh oh! Login screen unavailable!", buf);
                /*send_http_page(r, "很抱歉! 登陆页面存在问题!", buf);*/
		free(buf);
		debug(LOG_INFO, "Sent %s an apology since auth server not online - no point sending them to auth server", r->clientAddr);
	}
	else {
		/* Re-direct them to auth server */
		char *urlFragment;

		if (!(mac = arp_get(r->clientAddr))) {
			/* We could not get their MAC address */
			debug(LOG_INFO, "Failed to retrieve MAC address for ip %s, so not putting in the login request", r->clientAddr);
			safe_asprintf(&urlFragment, "%sgw_address=%s&gw_port=%d&gw_id=%s&url=%s&clientip=%s",
				auth_server->authserv_login_script_path_fragment,
				config->gw_address,
				config->gw_port, 
				config->gw_id,
				url,
				r->clientAddr);
		} else {			
			debug(LOG_INFO, "Got client MAC address for ip %s: %s", r->clientAddr, mac);
			safe_asprintf(&urlFragment, "%sgw_address=%s&gw_port=%d&gw_id=%s&mac=%s&url=%s&clientip=%s",
				auth_server->authserv_login_script_path_fragment,
				config->gw_address,
				config->gw_port, 
				config->gw_id,
				mac,
				url,
				r->clientAddr);
		}

		debug(LOG_INFO, "Captured %s requesting [%s] and re-directing them to login page", r->clientAddr, url);
		http_send_redirect_to_auth(r, urlFragment, "Redirect to login page");
		free(urlFragment);
	}
	free(url);
}

void 
http_callback_wifidog(httpd *webserver, request *r)
{
	send_http_page(r, "WiFiAS", "Please use the menu to navigate the features of this WiFiAS installation.");
}

void 
http_callback_about(httpd *webserver, request *r)
{
	send_http_page(r, "About WiFiAS", "This is WiFiDog version <strong>" VERSION "</strong>");
}

void 
http_callback_status(httpd *webserver, request *r)
{
	const s_config *config = config_get_config();
	char * status = NULL;
	char *buf;

	if (config->httpdusername && 
			(strcmp(config->httpdusername, r->request.authUser) ||
			 strcmp(config->httpdpassword, r->request.authPassword))) {
		debug(LOG_INFO, "Status page requested, forcing authentication");
		httpdForceAuthenticate(r, config->httpdrealm);
		return;
	}

	status = get_status_text();
	safe_asprintf(&buf, "<pre>%s</pre>", status);
	send_http_page(r, "WiFiAS Status", buf);
	free(buf);
	free(status);
}
/** @brief Convenience function to redirect the web browser to the auth server
 * @param r The request
 * @param urlFragment The end of the auth server URL to redirect to (the part after path)
 * @param text The text to include in the redirect header ant the mnual redirect title */
void http_send_redirect_to_auth(request *r, const char *urlFragment, const char *text)
{
	char *protocol = NULL;
	int port = 80;
	t_auth_serv	*auth_server = get_auth_server();

	if (auth_server->authserv_use_ssl) {
		protocol = "https";
		port = auth_server->authserv_ssl_port;
	} else {
		protocol = "http";
		port = auth_server->authserv_http_port;
	}
			    		
	char *url = NULL;
	safe_asprintf(&url, "%s://%s:%d%s%s",
		protocol,
		auth_server->authserv_hostname,
		port,
		auth_server->authserv_path,
		urlFragment
	);
	http_send_redirect(r, url, text);
	free(url);	
}

/** @brief Sends a redirect to the web browser 
 * @param r The request
 * @param url The url to redirect to
 * @param text The text to include in the redirect header and the manual redirect link title.  NULL is acceptable */
void http_send_redirect(request *r, const char *url, const char *text)
{
	char *message = NULL;
	char *header = NULL;
	char *response = NULL;
		/* Re-direct them to auth server */
	debug(LOG_DEBUG, "Redirecting client browser to %s", url);
	safe_asprintf(&header, "Location: %s", url);
	safe_asprintf(&response, "302 %s\n", text ? text : "Redirecting");
	httpdSetResponse(r, response);
	httpdAddHeader(r, header);
	free(response);
	free(header);
	safe_asprintf(&message, "Please<a href='%s'>Click here</a>.", url);
	send_http_page(r, text ? text : "Redirection to message", message);
	free(message);
}

void 
http_callback_auth(httpd *webserver, request *r)
{
	t_client	*client;
	httpVar * token;
	char	*mac;
	httpVar *logout = httpdGetVariableByName(r, "logout");
	if ((token = httpdGetVariableByName(r, "token"))) {
		/* They supplied variable "token" */
		if (!(mac = arp_get(r->clientAddr))) {
			/* We could not get their MAC address */
			debug(LOG_ERR, "Failed to retrieve MAC address for ip %s", r->clientAddr);
			send_http_page(r, "WiFiDog Error", "Failed to retrieve your MAC address");
		} else {
			/* We have their MAC address */

			LOCK_CLIENT_LIST();
			
			if ((client = client_list_find(r->clientAddr, mac)) == NULL) {
				debug(LOG_DEBUG, "New client for %s", r->clientAddr);
				client_list_append(r->clientAddr, mac, token->value);
			} else if (logout) {
			    t_authresponse  authresponse;
			    s_config *config = config_get_config();
			    unsigned long long incoming = client->counters.incoming;
			    unsigned long long outgoing = client->counters.outgoing;
			    char *ip = safe_strdup(client->ip);
			    char *urlFragment = NULL;
			    t_auth_serv	*auth_server = get_auth_server();
			    				    	
			    fw_deny(client->ip, client->mac, client->fw_connection_state);
			    client_list_delete(client);
			    debug(LOG_DEBUG, "Got logout from %s", client->ip);
			    
			    /* Advertise the logout if we have an auth server */
			    if (config->auth_servers != NULL) {
					UNLOCK_CLIENT_LIST();
					auth_server_request(&authresponse, REQUEST_TYPE_LOGOUT, ip, mac, token->value, 
									    incoming, outgoing);
					LOCK_CLIENT_LIST();
					
					/* Re-direct them to auth server */
					debug(LOG_INFO, "Got manual logout from client ip %s, mac %s, token %s"
					"- redirecting them to logout message", client->ip, client->mac, client->token);
					safe_asprintf(&urlFragment, "%smessage=%s",
						auth_server->authserv_msg_script_path_fragment,
						GATEWAY_MESSAGE_ACCOUNT_LOGGED_OUT
					);
					http_send_redirect_to_auth(r, urlFragment, "Redirect to logout message");
					free(urlFragment);
			    }
			    free(ip);
 			} 
 			else {
				debug(LOG_DEBUG, "Client for %s is already in the client list", client->ip);
			}
			UNLOCK_CLIENT_LIST();
			if (!logout) {
				authenticate_client(r);
			}
			free(mac);
		}
	} else {
		/* They did not supply variable "token" */
		send_http_page(r, "WiFiAS error", "Invalid token");
	}
}

void send_http_page(request *r, const char *title, const char* message)
{
    s_config	*config = config_get_config();
    char *buffer;
    struct stat stat_info;
    int fd;
    ssize_t written;

    fd=open(config->htmlmsgfile, O_RDONLY);
    if (fd==-1) {
        debug(LOG_CRIT, "Failed to open HTML message file %s: %s", config->htmlmsgfile, strerror(errno));
        return;
    }

    if (fstat(fd, &stat_info)==-1) {
        debug(LOG_CRIT, "Failed to stat HTML message file: %s", strerror(errno));
        close(fd);
        return;
    }

    buffer=(char*)safe_malloc(stat_info.st_size+1);
    written=read(fd, buffer, stat_info.st_size);
    if (written==-1) {
        debug(LOG_CRIT, "Failed to read HTML message file: %s", strerror(errno));
        free(buffer);
        close(fd);
        return;
    }
    close(fd);

    buffer[written]=0;
    httpdAddVariable(r, "title", title);
    httpdAddVariable(r, "message", message);
    httpdAddVariable(r, "nodeID", config->gw_id);
    httpdOutput(r, buffer);
    free(buffer);
}

